### ISSUE TITLE
Custom vibrations for MiBand

### ISSUE THREAD

**Issue by renno-bih** [Freeyourgadget/Gadgetbridge#402]
Hi, I was wondering if there is possibility to implement more options for generating custom vibrations for Xiaomi MiBand wristbands.  To be more specific, I need solution to generate only one vibration.
All current solutions are really buggy, there is no way produce reliable vibration.

If there could be solution to define duration and repetition of vibration it would be great,

**Comment by cpfeiffer** [comment]
What do you mean with current solutions? Gadgetbridge has support for some hardcoded [vibration profiles](https://github.com/Freeyourgadget/Gadgetbridge/blob/master/app/src/main/java/nodomain/freeyourgadget/gadgetbridge/devices/miband/VibrationProfile.java)

One could easily change that to add a single vibration. The number of repetitions is already configurable.

**Comment by renno-bih** [comment]
No matter what vibration or number of repetitions I choose I cannot get only one vibration. Two vibrations are minimum that I can produce.  That is what I tested on MiBand 1A (white LED version).

**Comment by renno-bih** [comment]
Tested with 1S, same problem.

**Comment by rplevka** [comment]
tested on my 1s and here's the output for various vibration profiles  and 1 repetition:
Settings for generic notifications and tested via debug->generic notification test

Profile | Vibration
-------- | ------------
Staccato | - -
Short | - or -.
Medium | --
Long | --
Waterdrop | -  or -.
Ring | -- --
Alarm Clock | -- -. or - -- or -. -- (seems pretty random)

Also, every now and then on updating the settings and testing the notification, i lost connection to band and had to restart the app to get connected again, not sure, why. Perhaps miband got rebooted?

**Comment by cpfeiffer** [comment]
The '.' is unfortunately out of our control. The only vibration supported by current firmware versions is '- -'. We try to cut off the second '-' by sending a "no vibration" command after a few ms, but the band sometimes honors it and sometimes not.

**Comment by cpfeiffer** [comment]
So it seems to be a Bluetooth latency problem. When I enable low latency mode, the vibration is a little better, and I definitely get just a single vibration (because our off command arrives in time).

Unfortunately entering low latency mode takes a little while, so the notification would be delayed by around 5-8s. And we cannot enable low latency mode in general because it consumes more battery.

**Comment by cpfeiffer** [comment]
AFAICS, this issue is more a request for a bugfix than an enhancement: make sure that the vibration profiles actually work as intended.

**Comment by renno-bih** [comment]
I'm wondering how they implemented vibrations in "Mi Band Control" application. 
They managed to implement short, medium and long vibrations and works great, any way to do reverse engineering and see how they managed it?

**Comment by Vebryn** [comment]
Using MiBand2, I feel no difference between vibration profiles!

**Comment by NotAFile** [comment]
Same, I don't feel any either. Is this intended (the mi2 simply can't do multiple patterns) or is this just a bug somewhere?

**Comment by ashimokawa** [comment]
I know how to do this on mi2, if someone wants to implement it i can tell how to do it.

**Comment by NotAFile** [comment]
Well, the knowledge won't get lost as long as you put it somewhere besides your head. I'm not really in the subject matter but if you share it I'm very sure someone will use the information.

**Comment by joserebelo** [comment]
@ashimokawa I would be willing to implement it, if you do say how :)

**Comment by Vebryn** [comment]
I tried with 0.26.5. Vibration patterns are now different. Issue solved!

**Comment by MarcoIeni** [comment]
I feel no difference with mi band 3

**Comment by avanc** [comment]
Just unpacked my Mi Band 3 and played with the vibration option. Unfortunately, there are also no differences. Just two short (and weak) vibrations. I'm afraid I won't feel them in the daily live.

**Comment by Alex2357** [comment]
Do anyone knows how to use this http://help.mibandtools.com/knowledge_base/topics/custom-notification-patterns ? I'd like to be able to send email to myself and control the vibrations for this email. Would it be possible?

**Comment by NotAFile** [comment]
Btw, I read that Xiaomi recently shipped a firmware update that allowed different vibration patterns natively.

**Comment by ashimokawa** [comment]
Yes I will try do do that. Bluetooth logs while running mi fit (and only mi fit) are welcome

**Comment by JPFrancoia** [comment]
I just got the Mi band 4, all the notifications are the same: 2 short vibrations. Changing settings on gadget bridge has no effect.
